libraries/chain/kv_context_rocksdb.cpp

```diff
@@ -0,0 +1,315 @@
+            actual_size = kv->key.size();
+            return kv_it_stat::iterator_ok;
+         } else {
+            actual_size = 0;
+            return kv_it_stat::iterator_end;
+         }
+      }
+
+      kv_it_stat kv_it_value(uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size) override {
+         EOS_ASSERT(!kv_it.is_erased(), kv_bad_iter, "Iterator to erased element");
+
+         std::optional<b1::chain_kv::key_value> kv;
+         try {
+            try {
+               kv = kv_it.get_kv();
+            }
+            FC_LOG_AND_RETHROW()
+         }
+         CATCH_AND_EXIT_DB_FAILURE()
+
+         if (kv) {
+            if (offset < kv->value.size())
+               memcpy(dest, kv->value.data() + offset, std::min((size_t)size, kv->value.size() - offset));
+            actual_size = kv->value.size();
+            return kv_it_stat::iterator_ok;
+         } else {
+            actual_size = 0;
+            return kv_it_stat::iterator_end;
+         }
+      }
+
+    private:
+      void get_current_key_value_sizes(uint32_t* found_key_size, uint32_t* found_value_size) {
+         if (kv_it.is_valid()) {
+            // Not end or at an erased kv pair
+            auto kv = kv_it.get_kv();
+
+            // kv is always non-null due to the check of is_valid()
+            *found_key_size = kv->key.size();
+            *found_value_size = kv->value.size();
+         } else {
+            *found_key_size = 0;
+            *found_value_size = 0;
+         }
+      }
+   }; // kv_iterator_rocksdb
+
+   struct kv_context_rocksdb : kv_context {
+      b1::chain_kv::database&                  database;
+      b1::chain_kv::undo_stack&                undo_stack;
+      name                                     database_id;
+      b1::chain_kv::write_session              write_session;
+      b1::chain_kv::view                       view;
+      name                                     receiver;
+      kv_resource_manager                      resource_manager;
+      const kv_database_config&                limits;
+      uint32_t                                 num_iterators = 0;
+      std::shared_ptr<const std::vector<char>> temp_data_buffer;
+
+      kv_context_rocksdb(b1::chain_kv::database& database, b1::chain_kv::undo_stack& undo_stack, name database_id,
+                         name receiver, kv_resource_manager resource_manager, const kv_database_config& limits)
+          : database{ database }, undo_stack{ undo_stack }, database_id{ database_id },
+            write_session{ database }, view{ write_session, make_prefix() }, receiver{ receiver },
+            resource_manager{ resource_manager }, limits{ limits } {}
+
+      ~kv_context_rocksdb() override {
+         try {
+            try {
+               write_session.write_changes(undo_stack);
+            }
+            FC_LOG_AND_RETHROW()
+         }
+         CATCH_AND_EXIT_DB_FAILURE()
```

Comment:
Are we sure we want to terminate the program?

Reply:
If this isn't a refactoring from the existing code, then we definitely cannot